dvc/tree/gdrive.py

```diff
@@ -510,35 +511,68 @@ def exists(self, path_info, use_dvcignore=True):
             return True
 
-    def _list_paths(self, prefix=None):
-        if not self._ids_cache["ids"]:
-            return
+    def ls(
+        self, path_info, detail=False, recursive=False
+    ):  # pylint: disable=arguments-differ
+        if recursive:
+            for entry in self.walk_files(path_info, detail=detail):
+                if detail:
+                    yield entry
+                else:
+                    yield str(entry)
+            return None
+
+        parent_ids = self._path_to_item_ids(path_info.path)
+        if not parent_ids:
+            return None
 
-        if prefix:
-            dir_ids = self._ids_cache["dirs"].get(prefix[:2])
-            if not dir_ids:
-                return
-        else:
-            dir_ids = self._ids_cache["ids"]
         parents_query = " or ".join(
-            f"'{dir_id}' in parents" for dir_id in dir_ids
+            f"'{dir_id}' in parents" for dir_id in parent_ids
         )
         query = f"({parents_query}) and trashed=false"
 
+        parent_path = path_info.path
         for item in self._gdrive_list(query):
-            parent_id = item["parents"][0]["id"]
-            yield posixpath.join(
-                self._ids_cache["ids"][parent_id], item["title"]
-            )
+            item_path = posixpath.join(parent_path, item["title"])
+            if detail:
+                if item["mimeType"] == FOLDER_MIME_TYPE:
+                    yield {"type": "directory", "name": item_path}
+                else:
+                    yield {
+                        "type": "file",
+                        "name": item_path,
+                        "size": item["fileSize"],
+                        "checksum": item["md5Checksum"],
+                    }
+            else:
+                yield item_path
+
+        if parent_ids:
+            self._cache_path_id(path_info.path, min(parent_ids))
+
+    def walk(self, path_info, detail=False):
+        stack = deque([path_info])
+        while stack:
+            path_info = stack.pop()
+            files, directories = [], []
+            for entry in self.ls(path_info, detail=True):
+                path = path_info.replace(path=entry["name"])
+                if entry["type"] == "file":
+                    if detail:
+                        files.append(entry)
+                    else:
+                        files.append(path)
+                else:
+                    directories.append(path)
+
+            yield path_info, directories, files
+            stack.extendleft(directories)
 
     def walk_files(self, path_info, **kwargs):
-        use_prefix = kwargs.pop("prefix", False)
-        if path_info == self.path_info or not use_prefix:
-            prefix = None
-        else:
-            prefix = path_info.path
-        for fname in self._list_paths(prefix=prefix, **kwargs):
-            yield path_info.replace(fname)
+        if kwargs.pop("prefix", False):
+            path_info = path_info / ""
```

Comment:
hmm, not sure I understand what's going on here ... could you please clarify? is it equivalent? do we change upstream a bit to pass path/prefix in a different way?

Reply:
I think this should be obsolete now